Update the_problem.md (#559)
cowprotocol/docs: docs/cow-protocol/reference/core/auctions/the_problem.md

Reviewer: This renders a bit strange as _surplus_function_. Maybe "... we define the surplus function ..."?

Author: I don't know if you can have a general definition of surplus function that could apply to any order (basket order or whatever). I guess beside saying that it should be increasing in the buy amounts and decreasing in the sell amounts, I don't know if you can say more *in general*. Of course, once you have a specific order type, you can specify the surplus function for that order type

Reviewer: I was trying to refer to the formatting of `... we define _surplus_function_ $$U_S:S\rightarrow \mathbb R$$, ...`. I see how in general surplus is not easy to define. One could leave out `_surplus_function_` or change it to `... the utility function ...`.

Reviewer: I still do not see the advantage of identifying orders with acceptance sets. For example there are orders with the same acceptance set but different protocol fees.

Author: Acceptance sets are defined from the users' perspective. It is fine if the orders have the same acceptance set but different fees: it just means that the "in/out net of fee" that a user is willing to accept is the same as that of another user

Reviewer: I see how the acceptance set is net of fees and a valid execution only depends on that set. But that set is not enough to encode all information of the order required for the mechanism.
So my gripe is more with concepts below which are defined in terms of the set but should be defined in terms of the order.

For example, two different orders can have the same acceptance set but different protocol fees. The protocol fee is not really an object `f_S` with an order `S`. Instead, there is an order `o` with an acceptance set `S_o` and a protocol fee function `f_o`.

None of that needs to be addressed in this PR though.

sidebar_position: 1
---

> # What is solving? A matehmatical, high-level description

Reviewer: The typo seems not to be fixed yet.

> # What is solving? A matehmatical, high-level description

Reviewer: ```suggestion
# What is solving? A mathematical, high-level description
```

Practically speaking, CoW Protocol allows only some types of orders, which we can think of as constraints on the set _S_ that a user can submit. One such constraint is that only pairwise swaps are allowed, that is, all vectors in $$S$$  have zeros in $$k-2$$ dimensions. Furthermore, each order must fit within one of the categories we now discuss. To simplify notation, when discussing these categories, we assume that $$k=2$$.

### Sell Orders

A  _sell order_ specifies a maximum sell amount of a given token _Y_ > 0, a buy token _b_, and a limit price $$\pi$$, that corresponds to the worst-case exchange rate that the user is willing to settle for. The limit price can be specified explicitly (as in the case of limit orders) or derived from an underlying quote and a slippage tolerance parameter (as in the case of market orders).

> Sell orders can be fill-or-kill whenever the executed sell amount must be Y (or nothing). They can be partially fillable if the executed sell amount can be smaller or equal to Y.  Formally, if _x_ denotes the (proposed) buy amount and _y_ denotes the (proposed) sell amount of the order, a fill-or-kill limit sell order has the form

Reviewer: This reads a bit confusing to me. Maybe something like
```suggestion
Sell orders can be fill-or-kill which means that the executed sell amount must be Y (or nothing). They can be partially fillable which means that the executed sell amount can be smaller or equal to Y.  Formally, if _x_ denotes the (proposed) buy amount and _y_ denotes the (proposed) sell amount of the order, a fill-or-kill limit sell order has the form
```
Or maybe "Sell orders are called fill-or-kill whenever the executed sell amount must be Y (or nothing). ..."

$$S=\left\{\begin{bmatrix} x \\-y \end{bmatrix}~~s.t. ~~\frac{y}{\pi}\leq x \hbox{ and } y\in\{0,Y\} \right\},$$

$$U(x,-y)= x-y / \pi$$.

The main difference is that the limit price $$\pi$$ corresponds to the price at which a zero-fee traditional AMM would trade. For example, in the case of a simple, constant product CoW AMM pool with reserves _X_ and _Y_ , we have

$$ \pi = X / (Y-y)$$

Finally, unlike sell and buy orders that are not valid anymore once executed, CoW AMM orders are always present. That is, as soon as a CoW AMM pool is created, a CoW AMM order for that pool is valid in all subsequent auctions. 

## Protocol Fees

Each user order may have an associated fee paid to the protocol. At a high level, these fees can be represented by a function that, for a given order $$S$$ maps all possible trades to a non-negative vector of tokens, that is $$f_S:S \rightarrow \mathbb R^k_+$$   with $$f_S(0)=0$$.

:::note

Solvers are also expected to charge a fee to cover the costs of executing an order (for example, gas). We discuss such fees later in the context of solvers' optimal bidding, but we do not account for them here as they are not part of the protocol.

:::

## Solution

> Solvers propose solutions to the protocol, where a solution is a set of trades to execute. Formally, suppose there are $$I$$ users and CoW AMM orders and _J_ are external liquidity sources. A solution is a list of trades $$\{o_1, o_2, ...o_I, l_1, l_2, ..., l_J\}$$ for each user, CoW AMM pool and external liquidity source such that:

Reviewer: ```suggestion
Solvers propose solutions to the protocol, where a solution is a set of trades to execute. Formally, suppose there are $$I$$ users and CoW AMM orders and _J_ external liquidity sources. A solution is a list of trades $$\{o_1, o_2, ...o_I, l_1, l_2, ..., l_J\}$$ for each user, CoW AMM pool and external liquidity source such that:
```

From the protocol viewpoint, each solution that satisfies the above constraints has a _score_ that is given by the total surplus generated and the fees paid to the protocol, all aggregated and denominated in some numéraire. More specifically, the score of a solution is equal to the sum of scores of the orders the solution proposes to execute, where the score of an order $$o$$ is defined as:

> * $$o$$ is a sell order or a CoW AMM order: $$\mathrm{score}(o) = (U(o)+ f(o)) \cdot p(b)$$, where $$p(b)$$ is an externally provided price of the buy token relative to a numéraire.

Reviewer: This is the intention, I guess, so we should keep this change.

But we should probably check internally if we want to enforce it. At the moment, CoW AMM orders can also be buy orders (and solvers can choose buy or sell orders based on whichever is best for them.

Author: right, I didn't know that. Then there are two types of CoW AMM orders: CoW AMM buy and CoW AMM sell orders

Reviewer: I think it is fine to only explain the sell order case for now, and we should encourage solvers to only use sell orders for consistency.
